src/convolution.jl

```diff
@@ -0,0 +1,91 @@
+"""
+    convolve(d1::T, d2::T) where T<:Distribution -> Distribution
+
+Convolve two distributions of the same type to yield the distribution corresponding to the
+sum of independent random variables drawn from the underlying distributions.
+
+The function is only defined in the cases where the convolution has a closed form as
+defined here https://en.wikipedia.org/wiki/List_of_convolutions_of_probability_distributions
+
+* `Bernoulli`
+* `Binomial`
+* `NegativeBinomial`
+* `Geometric`
+* `Poisson`
+* `Normal`
+* `Cauchy`
+* `Chisq`
+* `Exponential`
+* `Gamma`
+* `MultivariateNormal`
+"""
+
+# discrete univariate
+function convolve(d1::Bernoulli, d2::Bernoulli)
+    _check_convolution_args(d1.p, d2.p)
+    return Binomial(2, d1.p)
+end
+
+function convolve(d1::Binomial, d2::Binomial)
+    _check_convolution_args(d1.p, d2.p)
+    return Binomial(d1.n + d2.n, d1.p)
+end
+
+function convolve(d1::NegativeBinomial, d2::NegativeBinomial)
+    _check_convolution_args(d1.p, d2.p)
+    return NegativeBinomial(d1.r + d2.r, d1.p)
+end
+
+function convolve(d1::Geometric, d2::Geometric)
+    _check_convolution_args(d1.p, d2.p)
+    return NegativeBinomial(2, d1.p)
+end
+
+convolve(d1::Poisson, d2::Poisson) =  Poisson(d1.λ + d2.λ)
+
+
+# continuous univariate
+convolve(d1::Normal, d2::Normal) = Normal(d1.μ + d2.μ, hypot(d1.σ, d2.σ))
+convolve(d1::Cauchy, d2::Cauchy) = Cauchy(d1.μ + d2.μ, d1.σ + d2.σ)
+convolve(d1::Chisq, d2::Chisq) = Chisq(d1.ν + d2.ν)
+
+function convolve(d1::Exponential, d2::Exponential)
+    _check_convolution_args(d1.θ, d2.θ)
+    return Gamma(2, d1.θ)
+end
+
+function convolve(d1::Gamma, d2::Gamma)
+    _check_convolution_args(d1.θ, d2.θ)
+    return Gamma(d1.α + d2.α, d1.θ)
+end
+
+# continuous multivariate
+function convolve(
+    d1::Union{IsoNormal, ZeroMeanIsoNormal, DiagNormal, ZeroMeanDiagNormal},
+    d2::Union{IsoNormal, ZeroMeanIsoNormal, DiagNormal, ZeroMeanDiagNormal},
+    )
+    _check_convolution_shape(d1, d2)
+    return MvNormal(d1.μ .+ d2.μ, d1.Σ + d2.Σ)
+end
+
+function convolve(
+    d1::Union{FullNormal, ZeroMeanFullNormal},
+    d2::Union{FullNormal, ZeroMeanFullNormal},
+    )
+    _check_convolution_shape(d1, d2)
+    return MvNormal(d1.μ .+ d2.μ, d1.Σ.mat + d2.Σ.mat + 1e-12 * I)  # ensures positive definite
```

Comment:
else I get this error when convolving `fn1` with itself, which I'm guessing shouldn't happen...

```julia
 Expression: Matrix(d3.Σ) == Matrix(d1.Σ + d2.Σ)
  PosDefException: matrix is not positive definite; Cholesky factorization failed.
```

Reply:
If there's a better way to avoid it I'd be interested to hear

Reply:
can you post the matrix this happened with? I wonder if it is floating point round off? Mayeb need to wrap it in `Symmetric` or `Hermitian`

Reply:
```julia
julia> L1 = [1 0; 2 1]
2×2 Array{Int64,2}:
 1  0
 2  1

julia> fn1 = MvNormal(ones(2), cov(L1 * L1'))
FullNormal(
dim: 2
μ: [1.0, 1.0]
Σ: [0.5 1.5; 1.5 4.5]
)


julia> MvNormal(fn1.μ .+ fn1.μ, fn1.Σ + fn1.Σ)
ERROR: PosDefException: matrix is not positive definite; Cholesky factorization failed.
```

Reply:
Looking into it a bit more the first eigenvalue is practically 0 so this might be a bad example of a matrix to build a test on. But maybe warrants its own test.

What I can't understand is why `fn1.Σ` on its own is positive definite but `2 * fn1.Σ` isn't...
```julia
julia> eigen(fn1.Σ.mat)
2-element Array{Float64,1}:
 5.551115123125783e-17
 5.0   

julia> eigen(fn1.Σ.mat + fn1.Σ.mat)
2-element Array{Float64,1}:
  1.1102230246251565e-16
 10.0    

julia> eigen(fn1.Σ.mat + fn1.Σ.mat + 1e-12*I)
2-element Array{Float64,1}:
  1.000088900582341e-12
 10.000000000001   
```

Reply:
I think this is a weird floating point thing? adding/subtracting `1e-12*I` to `Σ` makes/stops `Σ` and `rΣ` being posdef (for r>0).

Reply:
I'd not do the ` + 1e-12 * I` here, and leave users to handle these weird edge cases... unless we think they're common?

Reply:
Well, albeit contrived, I've hit it on two distributions in my tests and few others from messing around locally. I wouldn't be against adding `+ eps(Float64) * I` (or maybe wrapping it in a try-catch with a warning to users) to ensure stability because the maths should allow self-convolution. But I'll defer to the opinion of others as I'm not an expert in this at all.

Reply:
I think `try ... catch ... + + eps(Float64) * I` is not too awful -- let's see what others think

Reply:
it would be better to avoid try-catching all over, it's slower. If something happens, better error right away, the other option is to have a keyword argument for a regularizing term to add, defaulting to `eps(T)`

Reply:
Note that
```

julia> [1;3]*[1 3]
2×2 Array{Int64,2}:
 1  3
 3  9
```
so this is not posdef and `fn1.Σ` is half of that, so also not posdef up to floating point errors. it is really not the right example. I don't really like adding a multiple of `I` though

Reply:
I've removed the regulariser from the function and instead used a better construction for the posdef matrices:

```julia
m = Symmetric(rand(2,2))
m2 = m^2
MvNormal(m2.data)
```

I'm not against a regulariser kwarg but I think it should be used in the wild first to determine how often it is needed.